File: LeetCode/0545.Boundary-of-Binary-Tree/Boundary-Of-Binary-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<Integer> res ;
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        res = new ArrayList<>();
        if(root == null) return res;
        res.add(root.val);
        leftBoundary(root.left);
        leaves(root.left);
        leaves(root.right);
        rightBoundary(root.right);
        return res;
    }
    public void leftBoundary(TreeNode root){
        if(root == null || (root.left == null && root.right == null) ){
            return;
        }
        res.add(root.val);
        if(root.left == null){
            leftBoundary(root.right);
        }
        else{
            leftBoundary(root.left);
        }
    }
    public void rightBoundary(TreeNode root){
        if(root == null || (root.left == null && root.right == null) ){
            return;
        }
        if(root.right == null){
            rightBoundary(root.left);
        }
        else{
            rightBoundary(root.right);
        }
        res.add(root.val);
        
    }
    public void leaves(TreeNode root){
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            res.add(root.val);
            return ;
        }
        leaves(root.left);
        leaves(root.right);
    }
}

File: LeetCode/0545.Boundary-of-Binary-Tree/Boundary-Of-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def boundaryOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root: return []
        res = []
        if root.left or root.right: res.append(root.val)
            
        node = root.left
        while node:
            if node.left or node.right: res.append(node.val)
            if node.left:
                node = node.left
            else:
                node = node.right
                
        stack = [root]
        while stack:
            node = stack.pop()
            if not node.left and not node.right:
                res.append(node.val)
            if node.right: stack.append(node.right)
            if node.left: stack.append(node.left)
                
        right = []
        node = root.right
        while node:
            if node.left or node.right: right = [node.val] + right
            if node.right:
                node = node.right
            else:
                node = node.left
        return res + right
